### PR TITLE
feat: add to_date function to StringValue

### PR DESCRIPTION
**Pull request by saschahofmann**
## Description of changes

* Adds `to_date` to string types that accepts a format string and parses a string to Date type.

* Uses a cast to timestamp and extract time as a fallback like this
```python
def visit_StringToDate(self, op, *, arg, format_str):
        return self.f.date(self.f.str_to_time(arg, format_str))
```
* Implements native functions for bigquery, clickhouse, MySQL, oracle, postgres, and snowflake

## Issues closed
Implements half of #8908

### CODE REVIEW

**Comment by saschahofmann** [comment]
@gforsyth I am struggling a little with running the tests. I couldn't find much on setting up the database. I can see that in conftest.py some env variable are read so I provided those for my local postgres database via the .env file (on a side note: I found it weird that the .env is committed? )

Now I seem to be missing an extension
```
psycopg2.errors.UndefinedFile: could not open extension control file "/opt/homebrew/share/postgresql@14/extension/plpython3u.control": No such file or directory
``` 

and it doesn't seem trivial to install it. Is there any documentation on setting up the env for running the tests?

**Comment by gforsyth** [comment]
Hey @saschahofmann -- rather than using your own instance, I'd recommend installing `just` and running `just up postgres`.
That will spin up a postgres container with all the extensions needed and at the port that our pytest setup expects by default.

**Comment by gforsyth** [comment]
There are also some instructions here: https://ibis-project.org/contribute/02_workflow#running-the-test-suite

If you're using a mac M2 it's more involved because of the architecture

**Comment by saschahofmann** [comment]
Perfect. No idea why i didnt find that section. Will experiment around and hopefully have something more complete soon.

**Comment by saschahofmann** [comment]
I know implemented the default fallback in sql/compiler.py L804 like this
```python
def visit_StringToDate(self, op, *, arg, format_str):
        return self.f.date(self.f.str_to_time(arg, format_str))
```
Is there a way to do it using ibis (something like `to_timestamp().date()`. Also I was thinking whether SQLGlot might have implemented it for its dialect already but I couldn't find it.

**Comment by saschahofmann** [comment]
For postgres, the format for to_timestamp seems to be translated to the postgres native format but its not obvious to me where this happens? I would need to do something similar for to_date

**Comment by gforsyth** [comment]
@saschahofmann -- `sqlglot` handles the timestamp formatting here: https://github.com/tobymao/sqlglot/blob/d3472c664fdfb7c9cfa9a54c6b0491b605cf4913/sqlglot/dialects/postgres.py#L186-L193

**Comment by saschahofmann** [comment]
Alright. I found that SQLGlot has actually implemented StrToDate so I am just using that. What do I need to do make the doc test pass?

File: ibis/expr/types/strings.py [hunk 1]
```diff
@@ -1302,6 +1302,35 @@ def to_timestamp(self, format_str: str) -> ir.TimestampValue:
...
+        >>> ibis.options.interactive = True
+        >>> t = ibis.memtable({"ts": ["20170206"]})
+        >>> t.ts.to_date("%Y%m%d")
+        ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+        ┃ StringToDate(ts, '%Y%m%d')      ┃
+        ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
+        │ date                            │
+        ├─────────────────────────────────┤
+        │ 2017-02-06                      │
+        └─────────────────────────────────┘
```
Comment (gforsyth) [inline ibis/expr/types/strings.py:1330]:
```suggestion
        ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
        ┃ StringToDate(ts, '%Y%m%d') ┃
        ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
        │ date                       │
        ├────────────────────────────┤
        │ 2017-02-06                 │
        └────────────────────────────┘
```

I think it's just an annoying terminal width issue -- this _should_ fix it

**Comment by saschahofmann** [comment]
I think its ready for review !

**Review by gforsyth — COMMENTED** [review]
This looks really nice, @saschahofmann !

I think this patch will get snowflake passing:

```
index 75bfd7ce6..c0915d059 100644
--- a/ibis/backends/snowflake/compiler.py
+++ b/ibis/backends/snowflake/compiler.py
@@ -78,6 +78,7 @@ class SnowflakeCompiler(SQLGlotCompiler):
         ops.Hash: "hash",
         ops.Median: "median",
         ops.Mode: "mode",
+        ops.StringToDate: "to_date",
         ops.StringToTimestamp: "to_timestamp_tz",
         ops.TimeFromHMS: "time_from_parts",
         ops.TimestampFromYMDHMS: "timestamp_from_parts",
```

**Comment by saschahofmann** [comment]
Huh, I think I actually had this but removed it because I thought it was in SQLGlot. Will put it back

**Comment by saschahofmann** [comment]
I assume that the failing test is unrelated? At least its not clear to me where this would be coming from?

**Review by gforsyth — APPROVED** [review]
Thanks for putting this together @saschahofmann!  The snowpark error is definitely unrelated and _might_ be flaky, so I'm going to merge this and see what happens on CI.

**Pull request merged**